toolchain/check/facet_type.cpp

```diff
@@ -263,4 +251,91 @@ auto AllocateFacetTypeImplWitness(Context& context,
   context.inst_blocks().ReplacePlaceholder(witness_id, empty_table);
 }
 
+auto IsPeriodSelf(Context& context, SemIR::ConstantId const_id) -> bool {
+  // This also rejects the singleton Error value as it's concrete.
+  if (!const_id.is_symbolic()) {
+    return false;
+  }
+  const auto& symbolic =
+      context.constant_values().GetSymbolicConstant(const_id);
+  // Fast early reject before doing more expensive operations.
+  if (symbolic.dependence != SemIR::ConstantDependence::PeriodSelf) {
+    return false;
+  }
+  return IsPeriodSelf(context, symbolic.inst_id);
+}
+
+auto IsPeriodSelf(Context& context, SemIR::InstId inst_id) -> bool {
+  // Unwrap the `FacetAccessType` instruction, which we get when the `.Self` is
+  // converted to `type`.
+  if (auto facet_access_type =
+          context.insts().TryGetAs<SemIR::FacetAccessType>(inst_id)) {
+    inst_id = facet_access_type->facet_value_inst_id;
+  }
+  if (auto bind_symbolic_name =
+          context.insts().TryGetAs<SemIR::BindSymbolicName>(inst_id)) {
+    const auto& bind_name =
+        context.entity_names().Get(bind_symbolic_name->entity_name_id);
+    return bind_name.name_id == SemIR::NameId::PeriodSelf;
+  }
+  return false;
+}
+
+auto ResolveRewriteConstraintsAndCanonicalize(Context& context,
+                                              SemIR::LocId loc_id,
+                                              SemIR::FacetTypeInfo& facet_type)
+    -> void {
+  // This operation sorts and dedupes the rewrite constraints. They are sorted
+  // primarily by the `lhs_id`, then by the `rhs_id`.
+  facet_type.Canonicalize();
+
+  if (facet_type.rewrite_constraints.empty()) {
+    return;
+  }
+
+  for (size_t i :
+       llvm::seq<size_t>(facet_type.rewrite_constraints.size() - 1)) {
+    auto& constraint = facet_type.rewrite_constraints[i];
+    if (constraint.lhs_id == SemIR::ErrorInst::InstId ||
+        constraint.rhs_id == SemIR::ErrorInst::InstId) {
+      continue;
+    }
+
+    if (!context.insts().Is<SemIR::ImplWitnessAccess>(constraint.lhs_id)) {
+      continue;
+    }
+    auto lhs_access =
+        context.insts().GetAs<SemIR::ImplWitnessAccess>(constraint.lhs_id);
+    if (!context.insts().Is<SemIR::LookupImplWitness>(lhs_access.witness_id)) {
+      continue;
+    }
+    auto lhs_lookup =
+        context.insts().GetAs<SemIR::LookupImplWitness>(lhs_access.witness_id);
+    if (!IsPeriodSelf(context, lhs_lookup.query_self_inst_id)) {
```

Comment:
Similar to the above:

```suggestion
    auto lhs_lookup =
        context.insts().TryGetAs<SemIR::LookupImplWitness>(lhs_access->witness_id);
    if (!lhs_lookup) {
      continue;
    }
    if (!IsPeriodSelf(context, lhs_lookup->query_self_inst_id)) {
```

Reply:
Done